api/pkg/auth/service/service.go

```diff
@@ -92,58 +92,126 @@ func createChecksum(token string) string {
 	return hex.EncodeToString(hash[:])
 }
 
-func (r *request) insertData(ghUser goth.User, code string) error {
+/*
+(Keeping email as the primary key in users table)
+We perform search query in users table on the basis of email:
+If Email is not present(existing user or new user)
+	query the accounts table on the basis of username and provider
+		- If the above case matches -> update the email in users table and other details respectively in accounts table
+		- If the above case returns empty record(new user)
+			- create a new user in the database
+If Email is already present
+	fetch the user_id from users table and query the accounts table with `where` clause of `user_id` and `provider`
+		- if the record doesn't exists(user trying to login with new provider)
+			then create a new record in accounts table with new provider and the user_id already associated with the email
+		- if record exists then update the record if there are any changes
+*/
+func (r *request) insertData(gitUser goth.User, code, provider string) error {
+
+	var acc model.Account
+	var user model.User
+
+	userQuery := r.db.Model(&model.User{}).
+		Where("email = ?", gitUser.Email)
 
 	// Check if user exist
-	q := r.db.Model(&model.User{}).
-		Where("github_login = ?", ghUser.NickName)
+	err := userQuery.First(&user).Error
 
-	var user model.User
-	err := q.First(&user).Error
+	// If email doesn't exists in users table
 	if err != nil {
+		// Check whether username and provider are matching
+		accountQuery := r.db.Model(&model.Account{}).Where("username = ?", gitUser.NickName).Where("provider = ?", provider)
+		err = accountQuery.First(&acc).Error
+
 		// If user doesn't exist, create a new record
 		if err == gorm.ErrRecordNotFound {
+			if user, err = r.insertIntoUsersTable(gitUser, code); err != nil {
+				r.log.Error(err)
+			}
 
-			user.GithubName = ghUser.Name
-			user.GithubLogin = ghUser.NickName
-			user.Type = model.NormalUserType
-			user.AvatarURL = ghUser.AvatarURL
-			user.Code = code
+			if err = r.insertIntoAccountsTable(gitUser, provider, user.ID); err != nil {
+				r.log.Error(err)
+				return err
+			}
+		} else {
+			// Account exists
+			// Update the user table with the email
+			if err := r.db.Model(&model.User{}).Where("id = ?", acc.UserID).
+				Updates(model.User{Code: code, Email: gitUser.Email, Type: model.NormalUserType}).Error; err != nil {
+				r.log.Error(err)
+				return err
+			}
+			// Update the AvatarUrl and Name in Accounts table
+			if err := accountQuery.Updates(model.Account{AvatarURL: gitUser.AvatarURL, Name: gitUser.Name}).Error; err != nil {
```

Comment:
So considering the scenario if user already exists, do we need to still update the avatar url and name
For example currently we have users logged in through github and we have user's name and avatar url, hence according to the condition we are again updating the avatar url and name. May be we can see if the provider is github then don't update the avatar url and name